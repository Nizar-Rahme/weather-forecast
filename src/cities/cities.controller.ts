import {
  Body,
  CacheInterceptor,
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Post,
  UseInterceptors,
} from '@nestjs/common';

import { CitiesService } from './cities.service';
import { CreateCityDto } from './dto/create-city.dto';

interface Weather {
  date: Date;
  isLowExceeded: boolean;
  minTemp: number;
}

interface City {
  name: string;
  country: string;
  lowTempLimit: number;
  forecast: Weather[];
}

@Controller({ path: 'v1/cities' })
@UseInterceptors(CacheInterceptor)
export class CitiesController {
  constructor(private citiesService: CitiesService) {}

  @Get()
  findAll(): Promise<City[]> {
    return this.citiesService.findAll();
  }

  @Get(':id')
  async findOne(@Param('id') id: string): Promise<City> {
    const city = await this.citiesService.findOne(id);

    if (city) return city;

    throw new HttpException('Not found', HttpStatus.NOT_FOUND);
  }

  @Post()
  async create(@Body() createCityDto: CreateCityDto) {
    await this.citiesService.createOrUpdate(createCityDto);
  }
}
