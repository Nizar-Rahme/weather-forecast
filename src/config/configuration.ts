export const configuration = () => ({
  API_KEY: process.env.API_KEY,
  CITIES: process.env.CITIES,
  INTERVAL: process.env.INTERVAL,
  MIN_TEMP: process.env.MIN_TEMP,
  MONGODB_URI: process.env.MONGODB_URI,
});
