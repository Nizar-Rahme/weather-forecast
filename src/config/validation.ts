import * as Joi from 'joi';

export const validationSchema = Joi.object({
  API_KEY: Joi.string().required(),
  API_URL: Joi.string().required(),
  CITIES: Joi.string()
    .pattern(/^(?:[\w\s]+,\w{2}\|)*[\w\s]+,\w{2}$/)
    .required(),
  INTERVAL: Joi.number().min(10000).default(120000),
  MIN_TEMP: Joi.number().required(),
  MONGODB_URI: Joi.string().required(),
  DB_USERNAME: Joi.string().required(),
  DB_PASS: Joi.string().required(),
});
