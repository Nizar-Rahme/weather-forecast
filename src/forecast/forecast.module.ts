import { HttpModule, Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';

import { ForecastService } from './forecast.service';
import { Forecast, ForecastSchema } from './schemas/forecast.schema';

@Module({
  imports: [
    HttpModule,
    MongooseModule.forFeature([
      { name: Forecast.name, schema: ForecastSchema },
    ]),
  ],
  providers: [ForecastService],
  exports: [ForecastService],
})
export class ForecastModule {}
