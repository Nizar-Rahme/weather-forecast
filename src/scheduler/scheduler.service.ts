import { Injectable } from '@nestjs/common';
import { SchedulerRegistry } from '@nestjs/schedule';

@Injectable()
export class SchedulerService {
  constructor(private readonly schedulerRegistry: SchedulerRegistry) {}

  addInterval(name: string, milliseconds: number, callback: () => void) {
    callback();
    const interval = setInterval(callback, milliseconds);
    this.schedulerRegistry.addInterval(name, interval);
  }
}
