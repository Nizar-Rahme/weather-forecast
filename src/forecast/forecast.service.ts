import { HttpService, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { InjectModel } from '@nestjs/mongoose';
import { isValidObjectId, Model, ObjectId } from 'mongoose';

import { CityDocument } from '../cities/schemas/city.schema';
import { Forecast, ForecastDocument } from './schemas/forecast.schema';
import { normalizeForecastData } from './utils/forecast.utils';

interface IForecast {
  city: {
    name: string;
    country: string;
  };
  list: [
    {
      main: {
        temp_min: number;
      };
      dt_txt: string;
    },
  ];
}

@Injectable()
export class ForecastService {
  constructor(
    private configService: ConfigService,
    private httpService: HttpService,
    @InjectModel(Forecast.name) private forecastModel: Model<ForecastDocument>,
  ) {}

  private readonly url = this.configService.get<string>('forecastAPI.url');
  private readonly key = this.configService.get<string>('forecastAPI.key');

  async fetchForecast({ name, country }: CityDocument) {
    try {
      const { data } = await this.httpService
        .get<IForecast>(this.url, {
          params: {
            units: 'metric',
            q: `${name},${country}`,
            appid: this.key,
          },
        })
        .toPromise();

      return normalizeForecastData(data);
    } catch (error) {
      return [];
    }
  }

  updateForecast(
    forecast: Pick<Forecast, 'date' | 'minTemp'>[],
    city: CityDocument,
  ) {
    return Promise.all(
      forecast.map(({ date, minTemp }) =>
        this.forecastModel
          .findOneAndUpdate(
            {
              date,
              city,
            },
            { minTemp },
            { upsert: true },
          )
          .exec(),
      ),
    );
  }

  async findByCityId(id: string | ObjectId): Promise<ForecastDocument[]> {
    if (!isValidObjectId(id)) return [];

    return this.forecastModel
      .where({ city: id })
      .sort({ date: 1 })
      .populate('city')
      .exec();
  }
}
