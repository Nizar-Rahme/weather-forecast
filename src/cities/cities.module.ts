import { CacheModule, Module } from '@nestjs/common';
import { CitiesController } from './cities.controller';
import { MongooseModule } from '@nestjs/mongoose';

import { ForecastModule } from '../forecast/forecast.module';
import { CitiesService } from './cities.service';
import { City, CitySchema } from './schemas/city.schema';
import { Forecast, ForecastSchema } from '../forecast/schemas/forecast.schema';
import { SchedulerModule } from '../scheduler/scheduler.module';

@Module({
  imports: [
    CacheModule.register(),
    MongooseModule.forFeature([
      { name: City.name, schema: CitySchema },
      { name: Forecast.name, schema: ForecastSchema },
    ]),
    SchedulerModule,
    ForecastModule,
  ],
  controllers: [CitiesController],
  providers: [CitiesService],
  exports: [CitiesService],
})
export class CitiesModule {}
