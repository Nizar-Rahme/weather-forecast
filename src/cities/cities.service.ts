import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { ConfigService } from '@nestjs/config';
import { Model, ObjectId } from 'mongoose';
import { Cache } from 'cache-manager';

import { ForecastService } from '../forecast/forecast.service';
import { City, CityDocument } from './schemas/city.schema';
import { CreateCityDto } from './dto/create-city.dto';
import { SchedulerService } from '../scheduler/scheduler.service';

interface Weather {
  date: Date;
  isLowExceeded: boolean;
  minTemp: number;
  updatedAt?: Date;
}

interface ICity {
  id: ObjectId;
  name: string;
  country: string;
  lowTempLimit: number;
  forecast: Weather[];
}

const CITIES_CACHE_KEY = 'CitiesService.Cities';

@Injectable()
export class CitiesService {
  private interval = this.configService.get<number>('config.interval');
  private lowTempLimit = this.configService.get<number>('config.lowTempLimit');
  private cities = this.configService.get<string[][]>('config.cities');

  constructor(
    private configService: ConfigService,
    private forecastService: ForecastService,
    private schedularService: SchedulerService,
    @Inject(CACHE_MANAGER) private cacheManager: Cache,
    @InjectModel(City.name) private cityModel: Model<CityDocument>,
  ) {
    this.getCitiesDocs();

    this.schedularService.addInterval(
      'ForecastService.fetchForecast',
      this.interval,
      async () => {
        const citiesDocs = await this.getCitiesDocs();

        citiesDocs.forEach(async (city) => {
          const forecast = await this.forecastService.fetchForecast(city);
          this.forecastService.updateForecast(forecast, city);
        });
      },
    );
  }

  private async getCitiesDocs() {
    const cashedCities = await this.cacheManager.get<CityDocument[]>(
      CITIES_CACHE_KEY,
    );

    if (cashedCities) return cashedCities;

    const citiesDocs = await Promise.all(
      this.cities.map(([name, country]) =>
        this.createOrUpdate({ name, country }),
      ),
    );

    return this.cacheManager.set<CityDocument[]>(CITIES_CACHE_KEY, citiesDocs, {
      ttl: null,
    });
  }

  createOrUpdate({ name, country }: CreateCityDto) {
    return this.cityModel
      .findOneAndUpdate(
        {
          name,
          country,
        },
        {},
        { upsert: true },
      )
      .exec();
  }

  async findAll(): Promise<ICity[]> {
    const cities = await this.getCitiesDocs();

    return Promise.all(
      cities.map(async ({ _id: id, name, country }) => {
        const forecastData = await this.forecastService.findByCityId(id);

        return {
          id,
          name,
          country,
          lowTempLimit: this.lowTempLimit,
          forecast: forecastData.map(({ date, minTemp }) => ({
            date,
            minTemp,
            isLowExceeded: minTemp < this.lowTempLimit,
          })),
        };
      }),
    );
  }

  async findOne(id: string): Promise<ICity | null> {
    const forecastData = await this.forecastService.findByCityId(id);

    if (forecastData.length === 0) return null;

    return {
      id: forecastData[0].city._id,
      name: forecastData[0].city.name,
      country: forecastData[0].city.country,
      lowTempLimit: this.lowTempLimit,
      forecast: forecastData.map(({ date, minTemp }) => ({
        date,
        minTemp,
        isLowExceeded: minTemp < this.lowTempLimit,
      })),
    };
  }
}
