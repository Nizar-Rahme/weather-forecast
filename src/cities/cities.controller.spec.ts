import { CacheModule, HttpException } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { ScheduleModule } from '@nestjs/schedule';
import { getModelToken } from '@nestjs/mongoose';
import { Types } from 'mongoose';

import { CitiesController } from './cities.controller';
import { CitiesService } from './cities.service';
import { ForecastService } from '../forecast/forecast.service';
import { SchedulerModule } from '../scheduler/scheduler.module';
import { City } from './schemas/city.schema';
import { City as ICity } from './interfaces/city.interface';
import { Forecast } from '../forecast/schemas/forecast.schema';

class ForecastModule {
  findByCityId(id: string) {
    return [];
  }
}

const mockCity: ICity = {
  id: new Types.ObjectId('603e45c0cf2c7cfe404cb0d3'),
  name: 'Helsinki',
  country: 'FI',
  lowTempLimit: 1,
  forecast: [
    {
      date: new Date(),
      minTemp: -3,
      isLowExceeded: true,
    },
  ],
};

describe('CitiesController', () => {
  let controller: CitiesController;
  let citiesService: CitiesService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot(),
        CacheModule.register(),
        ScheduleModule.forRoot(),
        SchedulerModule,
      ],
      providers: [
        { provide: ForecastService, useClass: ForecastModule },
        {
          provide: getModelToken(City.name),
          useValue: {},
        },
        {
          provide: getModelToken(Forecast.name),
          useValue: {},
        },
        CitiesService,
      ],
      controllers: [CitiesController],
    }).compile();

    controller = module.get<CitiesController>(CitiesController);
    citiesService = module.get<CitiesService>(CitiesService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('findAll', () => {
    it('should return an array of cities', async () => {
      expect.assertions(1);

      const result = [mockCity];
      jest
        .spyOn(citiesService, 'findAll')
        .mockImplementation(async () => result);

      expect(await controller.findAll(1)).toBe(result);
    });
  });

  describe('findOne', () => {
    it('should return the corresponding city', async () => {
      expect.assertions(1);

      const result = mockCity;

      jest
        .spyOn(citiesService, 'findOne')
        .mockImplementation(async () => result);

      expect(await controller.findOne('603e45c0cf2c7cfe404cb0d3', 1)).toBe(
        result,
      );
    });

    it('should throw an HttpException when the resource is not found', async () => {
      expect.assertions(1);

      jest.spyOn(citiesService, 'findOne').mockImplementation(async () => null);

      const findOne = () => controller.findOne('random', 1);

      await expect(findOne).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
