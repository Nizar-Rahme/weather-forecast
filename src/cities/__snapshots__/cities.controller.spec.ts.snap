// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesController findOne should throw an HttpException when the resource is not found 1`] = `"Not found"`;
